### PR TITLE
treewide: Update OSS projects revisions

### PR DESCRIPTION
**Pull request by tejlmand**
sdk-zephyr: https://github.com/nrfconnect/sdk-zephyr/pull/354
sdk-mcuboot: https://github.com/nrfconnect/sdk-mcuboot/pull/120
sdk-mcumgr: https://github.com/nrfconnect/sdk-mcumgr/pull/10
sdk-nrfxlib: https://github.com/nrfconnect/sdk-nrfxlib/pull/266

### CODE REVIEW

**Review by pdunaj — APPROVED** [review]

**Review by jhn-nordic — APPROVED** [review]

**Review by MarekPorwisz — APPROVED** [review]

**Review by trond-snekvik — APPROVED** [review]

**Review by hakonfam — CHANGES_REQUESTED** [review]
`ninja flash` flashes `zephyr.hex` when it should flash `merged.hex` looking in to it.

**Comment by hakonfam** [comment]
https://github.com/hakonfam/fw-nrfconnect-nrf-1/commit/7790c54374d61fbb671baa47b59f3d805130079f

Here is a fix for the issue with merged.hex not being flashed

**Review by sigvartmh — CHANGES_REQUESTED** [review]
To me all the applications using BSDLib halts. Either on their `POST_KERNEL` `bsdlib_init` or when calling `bsdlib_init`.﻿

**Comment by rugeGerritsen** [comment]
Github reports that CI is passing, but that does not seem to be the case. @tejlmand , are you looking into it?

**Comment by tejlmand** [comment]
> To me all the applications using BSDLib halts. Either on their `POST_KERNEL` `bsdlib_init` or when calling `bsdlib_init`.

I see some strange behavior if flashing using `ninja flash` or `west flash`.
I managed to successfully flash and run `asset_tracker` on nrf9160dk by manually doing:
- `nrfjprog --eraseall`
- `nrfjprog --program build/zephyr/merged.hex --verify`

@sigvartmh can you try to do the same, and let me know the result on your end ?

**Comment by tejlmand** [comment]
> @sigvartmh can you try to do the same, and let me know the result on your end ?

After writing previous message, asset_tracker now works, also when flashed with `west` or `ninja`.

**Comment by rlubos** [comment]
> To me all the applications using BSDLib halts. Either on their POST_KERNEL bsdlib_init or when calling bsdlib_init.﻿

I confirm, the bsdlib initialization stalls at `bsd_init()` call. I can't really debug properly given it's a closed library, but I've noticed in the sources that this function will wait for the IPC IRQ to update internal state variable. When debugging the IPC IRQ never hits. Did something changed upstream in terms of IRQ handling for nRF91?

**Comment by lemrey** [comment]
The stall is an evident symptom, but it seems the issue is more general; timeouts do not seem to be working in Non-Secure.

**Comment by tejlmand** [comment]
> The stall is an evident symptom, but it seems the issue is more general; timeouts do not seem to be working in Non-Secure.

Correct, seems to be related to SPM, as this also fails on ble samples on nrf53, that doesn't use bsdlib

**Comment by jhn-nordic** [comment]
Github is acting a bit weird on me. Sorry about that.

**Comment by hakonfam** [comment]
@sigvartmh  @tejlmand  here is a fix for the SPM issues which caused the bsdlib hang

https://github.com/tejlmand/fw-nrfconnect-nrf/pull/1

**Review by sigvartmh — APPROVED** [review]
`aws_fota` and `http_application_update` samples work.

`http_application_update` compiled with a warning
```
../src/main.c: In function 'dfu_button_init':
../src/main.c:143:8: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  143 |  gpiob = device_get_binding(DT_GPIO_LABEL(DT_ALIAS(sw0), gpios));
```

**Comment by tejlmand** [comment]
> 
> 
> Github reports that CI is passing, but that does not seem to be the case. @tejlmand , are you looking into it?

Additional PR against internal test repos has been created.
Changes are also needed internally with this upmerge and those internal PRs must go in,after which CI must re-run this PR to get green build.

**Review by pdunaj — CHANGES_REQUESTED** [review]
I have run a quick test on mouse/dongle. There seems to be some problem with setting conn params.
```
[00713290] <wrn> ble_conn_params: Cannot update conn parameters for peer 0x200042ac (err:-5)
[00716522] <wrn> bt_hci_core: opcode 0xfd02 status 0x0c
```
0x0c is `BT_HCI_ERR_CMD_DISALLOWED`. We invoke `set_conn_params`.
Hi @rugeGerritsen, do you know if there were any changes that would affect how it works?

I will also push one cleanup commit for changes done to our usb module. *done*

**Comment by pdunaj** [comment]
Logger seems to be malfunctioning (broken?). It does not report that messages are dropped if there is no space in segger buffer (there are simply no logs displayed). When I expanded the buffer I get missing logs but at some point I see that log is cut
```
[00000300] <inf> usb_hid: Device suspended
[00000300] <inf> usb_hid: Device suspended
[00000300] <wrn> usb_state: USB suspend
[00000301] <err> usb_state: PDUNAJ new_state 3
[00000318] <inf> bt_hci_

core: No ID address. App must call settings_load()
[00000318] <inf> ble_state: Bluetooth initialized
[00000327] <inf> ble_state: LLPM enabled
[00000331] <inf> event_manager: e: usb_state_event state:POWERED
```
there was 24k assigned for both logger and segger up buffer so there should be plenty of space.
Hi, @nordic-krch do you have any idea what may be wrong?

**Comment by pdunaj** [comment]
There are build issues in mcuboot
```
21/263] Building C object CMakeFiles/app.dir/main.c.obj
/home/pdunaj/work/ncs/bootloader/mcuboot/boot/zephyr/main.c: In function 'main':
/home/pdunaj/work/ncs/bootloader/mcuboot/boot/zephyr/main.c:352:17: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  352 |     detect_port = device_get_binding(CONFIG_BOOT_SERIAL_DETECT_PORT);
```
and more of similar.

**Comment by pdunaj** [comment]
nrf52820dongle_nrf52820 does not build - RAM overflow by 1.3k. It seems that Zephyr uses significantly more RAM now.

**Comment by wiminordic** [comment]
Why is this PR named so strangely? I think there was a naming convention agreed for manifest changes, they should contain "manifest" in the PR name.

**Comment by pdunaj** [comment]
There is one more issue here. The PR was named differently then the one done so far and the nrf_desktop tests were not executed. @wiminordic will update the filter but we can expect more issues to come.

**Comment by tejlmand** [comment]
> Why is this PR named so strangely? I think there was a naming convention agreed for manifest changes, they should contain "manifest" in the PR name.

Because it's an upmerge, similar to:
9ed32d39 treewide: Update OSS projects revisions
56a7dc83 treewide: Update OSS projects revisions
781d4814 treewide: synchronize upstream OSS projects
c02998d8 treewide: prefer CC310 to RNG + devicetree fixes

**Comment by tejlmand** [comment]
> There are build issues in mcuboot
> ...
> and more of similar.

because the references to the corresponding PRs in mcumgr and mcuboot accidentally was lost on a rebase.
Fixed.

**Comment by pdunaj** [comment]
> > Why is this PR named so strangely? I think there was a naming convention agreed for manifest changes, they should contain "manifest" in the PR name.
> 
> Because it's an upmerge, similar to:
> [9ed32d3](https://github.com/nrfconnect/sdk-nrf/commit/9ed32d3999fa776664182f30cd510cb9845ca5db) treewide: Update OSS projects revisions
> [56a7dc8](https://github.com/nrfconnect/sdk-nrf/commit/56a7dc8339fdcad8d53dad5a1691a1f0ac4c4ac7) treewide: Update OSS projects revisions
> [781d481](https://github.com/nrfconnect/sdk-nrf/commit/781d4814ac91b529bf5cef80e150baa87c8d5c2a) treewide: synchronize upstream OSS projects
> [c02998d](https://github.com/nrfconnect/sdk-nrf/commit/c02998d84899c48aa8c9de2ee78f52bee17fa1ef) treewide: prefer CC310 to RNG + devicetree fixes

I think this is simply misunderstanding. Since we don't want to run nrf_desktop tests all the time ages ago the filter was added to run them against PRs that have "manifest" or "nrf_desktop" in the title. Apparently at some point somebody named the upmerged PR differently, nobody noticed and so it went. The solution would be to trigger the full set of tests we have every time there is a change to "west.yml" as this is the only reliable way of telling if it is needed.
Generally nothing bad happened - we will just fix the issues found so far and run nrf_desktop tests when it's done.

**Comment by tejlmand** [comment]
> The solution would be to trigger the full set of tests we have every time there is a change to "west.yml" as this is the only reliable way of telling if it is needed.

All PRs touching the manifest will automatically be labelled with `manifest` label.
Maybe use that label for triggering nrf_desktop tests.

**Comment by pdunaj** [comment]
Going back to the size growth. I have run RAM report and the results are strange.

Competing new (upmerge PR) vs old (master)
RAM report says: 34985 vs 35131
build system says: 36058 vs 34762

So for master buildsystem thinks FW is smaller then report claims it to be while on upmerge it is the other way round...
I cannot seen any significant differences when comparing reports - there are changes both ways. But build system thinks new FW is larger by 1.3k and won't finish linking.

Hi @tejlmand, do you know what happened in build system or who keeps an eye on that?

**Comment by tejlmand** [comment]
> Hi @tejlmand, do you know what happened in build system or who keeps an eye on that?

A lot has happened in the build system itself, but nothing that should affect how things are reported to my knowledge.
Regarding increased ram usage, then we would need to look into actual modules to see what has changed.

File: drivers/gps/nrf9160_gps/nrf9160_gps.c
Comment (jhn-nordic) [inline drivers/gps/nrf9160_gps/nrf9160_gps.c:71]: Missing the const at the stop() function https://github.com/nrfconnect/sdk-nrf/blob/1f4578bf152f5d3c36442f6deeccb3e0a9fb83b8/drivers/gps/nrf9160_gps/nrf9160_gps.c#L724

**Review by jhn-nordic — CHANGES_REQUESTED** [review]

**Comment by jhn-nordic** [comment]
> There are build issues in mcuboot
> 
> ```
> 21/263] Building C object CMakeFiles/app.dir/main.c.obj
> /home/pdunaj/work/ncs/bootloader/mcuboot/boot/zephyr/main.c: In function 'main':
> /home/pdunaj/work/ncs/bootloader/mcuboot/boot/zephyr/main.c:352:17: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
>   352 |     detect_port = device_get_binding(CONFIG_BOOT_SERIAL_DETECT_PORT);
> ```
> 
> and more of similar.

Same seen when compiling for thingy91_nrf9160ns

**Comment by tejlmand** [comment]
> Same seen when compiling for thingy91_nrf9160ns

@jhn-nordic  Still see this ?
Or was it just the one I was already fixing and pushed before your comment.

**Comment by pdunaj** [comment]
Update from us:
- conn params setting is caused by configuration options change between host and LL
- our tests crash on assert in zephyr - it seems there is unhandled clock control event (`zephyr/drivers/clock_control/clock_control_nrf.c:599` r0 was 4 which is strange as it suggests audio clk not present on nrf52)
- the RAM increase issue is caused by constant orphan sections moved from flash to RAM; this is caused by `c8906fef79cefce3fe8c24cd0d01d164614d579f` but why it happens I really don't know and will have to look at it

**Comment by nordic-krch** [comment]
@pdunaj 
> our tests crash on assert in zephyr - it seems there is unhandled clock control event

I will investigate that.

**Comment by jhn-nordic** [comment]
> > Same seen when compiling for thingy91_nrf9160ns
> 
> @jhn-nordic Still see this ?
> Or was it just the one I was already fixing and pushed before your comment.

[47/250] Building C object zephyr/CMakeFiles/zephyr.dir/Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/serial_adapter.c.obj
/Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/serial_adapter.c: In function 'boot_uart_fifo_init':
/Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/serial_adapter.c:195:11: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
  uart_dev = device_get_binding(CONFIG_RECOVERY_UART_DEV_NAME);
           ^
/Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/serial_adapter.c:212:34: warning: passing argument 2 of 'uart_irq_callback_set' from incompatible pointer type [-Wincompatible-pointer-types]
  uart_irq_callback_set(uart_dev, boot_uart_fifo_callback);
                                  ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/serial_adapter.c:18:
/Users/jhn/work/repos/ncs/zephyr/include/drivers/uart.h:1131:37: note: expected 'uart_irq_callback_user_data_t' {aka 'void (*)(const struct device *, void *)'} but argument is of type 'void (*)(struct device *, void *)'
       uart_irq_callback_user_data_t cb)
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
[206/250] Building C object CMakeFiles/app.dir/main.c.obj
/Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/main.c: In function 'main':
/Users/jhn/work/repos/ncs/bootloader/mcuboot/boot/zephyr/main.c:352:17: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
     detect_port = device_get_binding(CONFIG_BOOT_SERIAL_DETECT_PORT);

**Review by rlubos — CHANGES_REQUESTED** [review]
nRF91 samples which use mcuboot (`lwm2m_client`, `asset_tracker`) don't boot (stall after `mcuboot: Jumping to the first image slot` message, tested with nRF9160dk).

Other nRF91 samples I've tested (coap, mqtt, https) work fine.

**Comment by tejlmand** [comment]
> Yes still get warnings when building asset tracker with: west build -b thingy91_nrf9160ns

cherry-picked: https://github.com/zephyrproject-rtos/mcuboot/pull/33

**Comment by tejlmand** [comment]
> nRF91 samples which use mcuboot (`lwm2m_client`, `asset_tracker`) don't boot (stall after `mcuboot: Jumping to the first image slot` message, tested with nRF9160dk).

we had similar issue which were fixed, and those samples should be running.
https://github.com/nrfconnect/sdk-nrf/pull/2918#issuecomment-696709083
@hakonfam could this be related ?

**Comment by hakonfam** [comment]
> > nRF91 samples which use mcuboot (`lwm2m_client`, `asset_tracker`) don't boot (stall after `mcuboot: Jumping to the first image slot` message, tested with nRF9160dk).
> 
> we had similar issue which were fixed, and those samples should be running.
> [#2918 (comment)](https://github.com/nrfconnect/sdk-nrf/pull/2918#issuecomment-696709083)
> @hakonfam could this be related ?

Yes, we saw the same behavior (bsdlib init hanging during system init - before main). But the fix resolved those issues.

If you break at `__start` in the application, do you see that execution continues there?

**Comment by rlubos** [comment]
> > > nRF91 samples which use mcuboot (`lwm2m_client`, `asset_tracker`) don't boot (stall after `mcuboot: Jumping to the first image slot` message, tested with nRF9160dk).
> > 
> > 
> > we had similar issue which were fixed, and those samples should be running.
> > [#2918 (comment)](https://github.com/nrfconnect/sdk-nrf/pull/2918#issuecomment-696709083)
> > @hakonfam could this be related ?
> 
> Yes, we saw the same behavior (bsdlib init hanging during system init - before main). But the fix resolved those issues.
> 
> If you break at `__start` in the application, do you see that execution continues there?

This bsdlib issue mentioned is fixed indeed - samples that do not use mcuboot work fine. This seems like a different one.

Debugging with `asset_tracker`, I don't even reach `__start` in the main app. When debugging `spm` image, I see it stops in the entry point, but I can only debug a few steps until I lose control of the execution.

**Comment by tejlmand** [comment]
> Debugging with `asset_tracker`, I don't even reach `__start` in the main app. When debugging `spm` image, I see it stops in the entry point, but I can only debug a few steps until I lose control of the execution.

I see the same, so something must have happened between the pre-upmerge, and the Zephyr 2.4 release.

**Comment by hakonfam** [comment]
> > Debugging with `asset_tracker`, I don't even reach `__start` in the main app. When debugging `spm` image, I see it stops in the entry point, but I can only debug a few steps until I lose control of the execution.
> 
> I see the same, so something must have happened between the pre-upmerge, and the Zephyr 2.4 release.

I have a fix.  It needs to go in mcuboot. 

```
diff --git a/boot/zephyr/main.c b/boot/zephyr/main.c
index 0eba4d7..cfff448 100644
--- a/boot/zephyr/main.c
+++ b/boot/zephyr/main.c
@@ -429,6 +429,9 @@ void main(void)
 
     ZEPHYR_BOOT_LOG_STOP();
 
+    __set_PSPLIM(0);
+    __set_MSPLIM(0);
+
     do_boot(&rsp);
 
     BOOT_LOG_ERR("Never should get here");
```

**Comment by pdunaj** [comment]
@tejlmand I found a workaround to deal with increased RAM usage.
I think linker places event manager data along with "devices" section (probably because this section's elements now have read only attribute). The problem is that some things in this section are still placed into RAM. Since these are the ones at the end, linker glues orphaned section behind it in RAM. At least this is what I think is happening after playing around a bit.
It seems to be enough to create additional section with element with attributes used by our orphan sections and linker can deal with placement now. So I added one along with event manager.
I created PR to master. https://github.com/nrfconnect/sdk-nrf/pull/3015

**Review by rlubos — APPROVED** [review]
The mcuboot/lwm2m_client issue was fixed, LGTM.

**Review by jtguggedal — APPROVED** [review]

**Review by ru-fu — APPROVED** [review]
No relevant doc changes.

**Review by jhn-nordic — APPROVED** [review]

**Review by pdunaj — APPROVED** [review]
run a quick smoke test with mouse and dongle - seems to be working fine

**Review by MarekPieta — CHANGES_REQUESTED** [review]
For nrf_desktop: ZDebugMCUBootSMP configuration does not work on nrf52840DK (application does not boot)

**Comment by tejlmand** [comment]
> For nrf_desktop: ZDebugMCUBootSMP configuration does not work on nrf52840DK (application does not boot)

@MarekPieta  could that be related to this ?
https://github.com/nrfconnect/sdk-nrf/pull/3028/files

**Comment by pdunaj** [comment]
> DebugMCUBootSMP configuration does not work

I checked `DebugMCUBootQSPI` and it boots. This using MCUBoot too. Let me check the other one.

**Comment by pdunaj** [comment]
> DebugMCUBootSMP configuration does not work

Yes, it seems to be unbootable.

```
*** Booting Zephyr OS build zephyr-v2.4.0-197-g86822c567484  ***
[00008480] <inf> mcuboot: Starting bootloader
[00008640] <inf> mcuboot: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[00008929] <inf> mcuboot: Boot source: none
[00009065] <inf> mcuboot: Swap type: none
[00021238] <inf> mcuboot: Bootloader chainload address offset: 0xc000
```

**Comment by MarekPieta** [comment]
@tejlmand, I think not. The MCUBoot image manager was always enabled by default.

**Comment by pdunaj** [comment]
It actually boots but it crashes when zephyr sets up the bss in application. Fault happens in the memset:
```
z_bss_zero () at /home/pdunaj/work/ncs/zephyr/kernel/init.c:108
(gdb) p __bss_end
$1 = 0x20006800 <logging_stack> "4h"
(gdb) p __bss_start
$2 = 0x20000e98 ""
```
I don't see anything suspicious in map. I will split the memset into several steps and see which range crash.

**Comment by pdunaj** [comment]
I recall - we now have stack guard enabled by default for all boards :D

**Comment by pdunaj** [comment]
I have disabled stack protection in our app for mcuboot. This will have to do until problem is solved.

**Review by MarekPieta — APPROVED** [review]
The issue related to nrf_desktop configuration with MCUBoot and SMP is fixed. The firmware is properly updated through the SMP.
No more issues were found.

**Comment by maciekfabia** [comment]
> It helps. When I disabled `CONFIG_HW_STACK_PROTECTION` application boots.

The same issue in Zigbee sample suing MCUBoot. Disabling `CONFIG_HW_STACK_PROTECTION` in MCUBoot helps

**Comment by pdunaj** [comment]
> > It helps. When I disabled `CONFIG_HW_STACK_PROTECTION` application boots.
> 
> The same issue in Zigbee sample suing MCUBoot. Disabling `CONFIG_HW_STACK_PROTECTION` in MCUBoot helps

It seems that the best thing will be to disable stack protection in default mcuboot configuration (that is used by samples).

**Comment by tejlmand** [comment]
> 
> 
> > > It helps. When I disabled `CONFIG_HW_STACK_PROTECTION` application boots.
> > 
> > 
> > The same issue in Zigbee sample suing MCUBoot. Disabling `CONFIG_HW_STACK_PROTECTION` in MCUBoot helps
> 
> It seems that the best thing will be to disable stack protection in default mcuboot configuration (that is used by samples).

but if mcuboot is built as part of multiimage, then this should be included:
https://github.com/nrfconnect/sdk-nrf/blob/ee9d25a75764811fdb638338794c510e1fd92dd2/subsys/bootloader/image/multi_image_mcuboot.conf#L1

**Comment by tejlmand** [comment]
has cherry-picked: https://github.com/JuulLabs-OSS/mcuboot/pull/828

**Comment by maciekfabia** [comment]
> has cherry-picked: [JuulLabs-OSS/mcuboot#828](https://github.com/JuulLabs-OSS/mcuboot/pull/828)

It solves issue in Zigbee. Edits related to `CONFIG_HW_STACK_PROTECTION` are not needed any more

**Review by nordic-auko — APPROVED** [review]
Connectivity bridge tested and works fine

**Comment by pdunaj** [comment]
@tejlmand CI is green :) this is our chance to have it in...

**Comment by mbolivar-nordic** [comment]
Dependent PRs are merged and west.yml is all SHAs again. Waiting for final CI run before merging.

**Comment by mbolivar-nordic** [comment]
@tejlmand CI is red now :cry:

**Comment by utzig** [comment]
Documentation build failed because of a key change in `sample.yaml`, I submitted a fix here: https://github.com/nrfconnect/sdk-nrf/pull/3052. I think this PR can be merged. Btw, the documentation publishing is currently disabled to avoid publishing broken documentation.

**Review by utzig — APPROVED** [review]

**Comment by utzig** [comment]
> Documentation build failed because of a key change in `sample.yaml`, I submitted a fix here: #3052. I think this PR can be merged. Btw, the documentation publishing is currently disabled to avoid publishing broken documentation.

Or otherwise just cherry-pick that commit!

**Comment by mbolivar-nordic** [comment]
> > 
> 
> Or otherwise just cherry-pick that commit!

It doesn't apply cleanly

**Comment by mbolivar-nordic** [comment]
@utzig can you fix the conflicts and push your commit on top of this PR?

**Comment by utzig** [comment]
> @utzig can you fix the conflicts and push your commit on top of this PR?

Oh, my bad, it's because the PR that introduced this was merged just one hour ago, then this one needs to be rebased on master again and then do the cherry-pick, sorry for that.

**Comment by utzig** [comment]
@tejlmand it's so beautiful when it's green!

**Comment by mbolivar-nordic** [comment]
Thanks @utzig !

**Pull request merged**
